File: python/239.py
class Solution(object):
	def maxSlidingWindow(self, nums, k):
		if not nums: return []

		window_index = []
		res = []
		for i, x in enumerate(nums):
			if i >= k and windows_index[0] <= i - k:
				window_index.pop(0)
			while window_index and nums[window_index[-1]] <= x:
				window_index.pop()

			window_index.append(i)

			if i >= k - 1:
				res.append(nums[window_index[0]])
		return res

File: java/239.java
public class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
    	if (nums == null || nums.length == 0 ) return new int[0];

		Queue<Integer> pq = new PriorityQueue<Integer>(k ,Collections.reverseOrder());

		int[] result = new int[nums.length-k+1];

		for (int i = 0;i < nums.length;++i){
			if (i >= k) {
				pq.remove(nums[i-k]);
			}
			pq.add(nums[i]);
			if (i >= k -1) {
				result[i - (k - 1)] = pq.peek();
			}
		}
		return result;
	}
}